Include/slp_structs.h

```diff
@@ -134,6 +134,7 @@ typedef struct _tasklet {
     int recursion_depth;
     PyObject *def_globals;
     PyObject *tsk_weakreflist;
+    PyObject *context;  /* if running: the saved context, otherwise the context for the tasklet */
```

Comment:
What is a _saved_ context?  If it is the context of a previous tasklet, surely that tasklet's context is stored in that tasklet's 
`context` member.

Reply:
In C-Python the thread state (`PyThreadState` or short `ts` for the current thread state) contains a counted reference to a Context object: `ts->context`. This reference may be NULL.

In Stackless Python each tasklet has its own counted reference to a Context object: `tasklet->context`. 

Now, if you switch from one tasklet to another the reference count of the involved context objects **must not change**. (During a tasklet switch a recursion into the interpreter is not allowed and therefore Py_DECREF is not allowed.) Therefore its is only possible to "exchange" context between variables.
1. exchange the values of old_tasklet->context and ts->context
2. exchange the values of new_tasklet->context and ts->context

If you create the main tasklet in `slp_initialize_main_and_current()` only one exchange happens:
1. exchange the values of main_tasklet->context and ts->context

Same on exit of the main tasklet in `schedule_task_destruct()`.
1. exchange the values of main_tasklet->context and ts->context

The effect of this schema is, that the value of tasklet->context depends on the state of the tasklet.
* if tasklet is current: tasklet->context is the value of ts->context before the main tasklet was created. ts->context is the context of the current tasklet.
* if tasklet is not current: tasklet->context is the context of the tasklet.

To answer your question: the "saved" context is the value of `ts->context` when the main-tasklet was created.
Of course I have to improve the comment.

Reply:
You are perfectly right. Keeping the context of the main thread is just a side effect of the implementation. And the implementation - the exchange values between thread-state and tasklet logic - was just an attempt to keep the reference counting correct without too much thinking about it: the exchange operation simply does not affect reference counts and there was already suitable macro/inline-function (`SLP_EXCHANGE_EXCINFO`). But your suggestion most likely also works an is much simpler. I'll give it a try.